python/lsst/sphgeom/_compoundRegion.cc

```diff
@@ -41,38 +41,60 @@ namespace sphgeom {
 
 namespace {
 
-py::str _repr(const char *format, CompoundRegion const &self) {
-    py::object first = py::cast(self.getOperand(0), py::return_value_policy::reference);
-    py::object second = py::cast(self.getOperand(1), py::return_value_policy::reference);
-    return py::str(format).format(first, second);
+py::str _repr(const char *class_name, CompoundRegion const &self) {
+    std::string format = class_name;
+    format += "(";
+    py::tuple operands(self.nOperands());
+    for (unsigned i = 0; i != self.nOperands(); i++) {
+        py::object operand = py::cast(self.getOperand(i), py::return_value_policy::reference);
+        operands[i] = operand;
+        if (i != 0) {
+            format += ", ";
+        }
+        format += "{!r}";
+    }
+    format += ")";
+    return py::str(format).format(*operands);
+}
+
+template <typename _CompoundRegion>
+std::unique_ptr<_CompoundRegion> _args_factory(const py::args& args) {
+    std::vector<std::unique_ptr<Region>> operands;
+    for (auto&& item: args) {
+        Region* region = item.cast<Region*>();
+        operands.emplace_back(region->clone());
+    }
+    return std::make_unique<_CompoundRegion>(std::move(operands));
 }
 
 }  // namespace
 
 template <>
 void defineClass(py::class_<CompoundRegion, std::unique_ptr<CompoundRegion>, Region> &cls) {
+    cls.def("nOperands", &CompoundRegion::nOperands);
     cls.def(
         "cloneOperand",
         [](CompoundRegion const &self, std::ptrdiff_t n) {
-            return self.getOperand(python::convertIndex(2, n)).clone();
+            int nOperands = self.nOperands();
+            return self.getOperand(python::convertIndex(nOperands, n)).clone();
```

Comment:
I'd love to have a way to let Python code iterate over regions without having to clone them, if we can do it safely.  As long as we don't provide a way to modify a compound region after it has been constructed, I believe pybind11's `return_value_policy::reference_internal` should make it safe to return a reference to an operand.

Reply:
I have added `__len__` and `__iter__` to CompoundRegion using `reference_internal` for latter. I did not make `_operands` cost, as on another commit I added method to flatten it in place after construction.